### PR TITLE
[SPARK-25392][Spark Job History]Inconsistent behaviour for pool details in spark web UI and history server page 

### PR DESCRIPTION
**Pull request by sandeep-katta**
## What changes were proposed in this pull request?

1.Added the new SparkListenerEvent **SparkListenerPoolInformation** which holds the pool Information as Map of (poolName -> Pool)

2.AppStatusListener reads this events and stores in kvstore

3.EventLoggingListener overrides the onPoolInformationEvent API and logs the event in Json Format,so History server can replay these events on start up

4, PoolInformation is visible in two pages: one is AllStagesPage, and the other is PoolPage,both will fetch the pool Details from kvstore and display accordingly

In summary, added the new ListenerEvent SparkListenerPoolInformation which holds the pool information, AppStatusListener writes this event to kvstore and AllStagesPage, Poolpage gets details from kvstore and display the information accordingly 

## How was this patch tested?

Added Testcases and also verified manually in the cluster

### CODE REVIEW

**Comment by sandeep-katta** [comment]
cc @vanzin @srowen ,Please review this improvement in History UI

**Comment by srowen** [comment]
I'm not sure I know enough to evaluate this. Is the solution to send all this info around, or just not link to a pool in this case? I don't know.

For some reason the Github diff is highlighting strings in dark red, as if there's a problem. I don't see one. Not sure why it is doing that.

**Comment by sandeep-katta** [comment]
Yes I am sending the pool Info around,so history server also can have details.

And regarding the dark red ,I think git unable to parse it correctly.

**Comment by sandeep-katta** [comment]
cc @cloud-fan Please review

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2434,8 +2434,15 @@ class SparkContext(config: SparkConf) extends Logging {
       val schedulingMode = getSchedulingMode.toString
       val addedJarPaths = addedJars.keys.toSeq
       val addedFilePaths = addedFiles.keys.toSeq
+      // SPARK-25392 pool Information should be stored in the event
+      val poolInformation = getAllPools.map { it =>
+        val xmlString = ("<pool><item PoolName=\"%s\" MinimumShare=\"%d\"" +
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Hmm, really don't like this kind of ad-hoc serialization format.

Instead, create a proper pool class for the REST API, and use that class to transfer pool information around.

Because this is a legacy event (not serialized by Jackson) you'll need some custom serialization in `JsonProtocol.scala`, but that's cleaner than what you have in this change.

Comment (sandeep-katta) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: cc @vanzin Have changed the code accordingly, can you please have a look at this

**Comment by cloud-fan** [comment]
I don't know this part well, and I have no idea why this problem occurs. Would be great to see more details/analysis.
